### ISSUE TITLE
Shap and interaction values differ with and without gpu

### ISSUE THREAD

**Issue by mahdi-ghelichi** [dmlc/xgboost#6069]
Shap and interaction values with and without gpu are different on xgboost.

**Comment by hcho3** [comment]
@mahdi-ghelichi Can you post an example where this is the case?

**Comment by mahdi-ghelichi** [comment]
from sklearn.datasets import load_iris

import xgboost as xgb

parameters = {"booster": "gbtree",

              "eval_metric": 'logloss',

              "eta": 0.5,

              "max_depth": 5,

              "nthread": 8,

              "objective": "binary:logistic",

              "seed": 42,

             "tree_method": "hist"}

 

if __name__=="__main__":

    iris = load_iris()

    data = iris['data']

    label = iris['target']

    label[label==2]=0

    train = xgb.DMatrix(data=data, label=label)

    model_no_gpu = xgb.train(parameters, train, num_boost_round=10)

 

    shap_values_no_gpu = model_no_gpu.predict(train, pred_contribs=True, approx_contribs=True)

    interactions_no_gpu = model_no_gpu.predict(train, pred_interactions=True)

 

    parameters['tree_method'] = 'gpu_hist'

    model_gpu = xgb.train(parameters, train, num_boost_round=10)

    shap_values_gpu = model_gpu.predict(train, pred_contribs=True, approx_contribs=True)

    interactions_gpu = model_gpu.predict(train, pred_interactions=True)

You will see that both shaps and interaction are different.

**Comment by hcho3** [comment]
Which version of XGBoost are you using? GPU-enabled SHAP was introduced recently and is not part of 1.2.0 release.

**Comment by mahdi-ghelichi** [comment]
It is 1.2.0 which is the latest.
So, it means till then we should not use shaps with gpu then.

Was wondering if we could get interactions with approx method? It returns all zero for other features when approx is True for interactions, except the diagonal. The approximation of interactions could speed things up.
